client/components/command-palette/use-command-palette.tsx

```diff
@@ -74,73 +75,94 @@ const siteToAction =
-		} );
-
-	// Create a variable to hold the final result
-	const finalSortedCommands = [ ...sortedCommands ];
+			// Map filtered sites to actions using the onClick function
+			sitesToPick = filteredSites.map( siteToAction( onClick ) );
+		}
 
-	// Add the "viewMySites" command to the beginning in all contexts except "/sites"
-	if ( viewMySitesCommand && currentPath !== '/sites' ) {
-		finalSortedCommands.unshift( viewMySitesCommand );
+		return { commands: sitesToPick ?? [] };
 	}
 
-	const selectedCommand = finalSortedCommands.find( ( c ) => c.name === selectedCommandName );
-	let sitesToPick = null;
-	if ( selectedCommand?.siteFunctions ) {
-		const { onClick, filter } = selectedCommand.siteFunctions;
-		let filteredSites = filter ? sortedSites.filter( filter ) : sortedSites;
-		if ( currentSiteId ) {
-			const currentSite = filteredSites.find( ( site ) => site.ID === currentSiteId );
-			if ( currentSite ) {
-				filteredSites = [
-					currentSite,
-					...filteredSites.filter( ( site ) => site.ID !== currentSiteId ),
-				];
-			}
+	// Logic for root commands
+	let finalSortedCommands = [];
+	if ( search ) {
+		const searchLowerCase = search.toLowerCase();
+		// Filter commands based on label and searchLabel
+		const commandLabelMatch = commands.filter( ( command ) =>
+			command.label.toLowerCase().includes( searchLowerCase )
+		);
+		const commandSearchLabelMatch = commands.filter(
+			( command ) =>
+				command.searchLabel?.toLowerCase().includes( searchLowerCase ) &&
+				! commandLabelMatch.includes( command )
+		);
+
+		// Combine label and searchLabel matches
+		finalSortedCommands = [ ...commandLabelMatch, ...commandSearchLabelMatch ];
+	} else {
+		// Filter out commands that have context
+		const commandHasContext = ( paths: string[] = [] ): boolean =>
+			paths.some( ( path ) => currentPath.includes( path ) ) ?? false;
+
+		const viewMySitesCommand = commands.find( ( command ) => command.name === 'viewMySites' );
+
+		// Sort commands with contextual commands ranking higher than general in a given context
+		const sortedCommands = commands
+			.filter( ( command ) => ! ( command === viewMySitesCommand ) )
+			.sort( ( a, b ) => {
+				const hasContextCommand = commandHasContext( a.context );
+				const hasNoContext = commandHasContext( b.context );
+
+				if ( hasContextCommand && ! hasNoContext ) {
+					return -1; // commands with context come first if there is a context match
+				} else if ( ! hasContextCommand && hasNoContext ) {
+					return 1; // commands without context set
+				}
+
+				return 0; // no change in order
+			} );
+
+		// Initialize the final result with sorted commands
+		finalSortedCommands = [ ...sortedCommands ];
+
+		// Add the "viewMySites" command to the beginning in all contexts except "/sites"
+		if ( viewMySitesCommand && currentPath !== '/sites' ) {
+			finalSortedCommands.unshift( viewMySitesCommand );
 		}
-		sitesToPick = filteredSites.map( siteToAction( onClick ) );
 	}
```

Comment:
It'd be great to have some tests around all of this sorting/filtering behavior. It's hard to assess whether fixing one scenario is breaking another scenario.

Reply:
100%, I was thinking about this already - @sejas I don't know how fast you would like to merge. Should we do a separate issue to add the tests for both sorting and filtering or should we handle it as a part of this PR?

Reply:
Let's handle the tests as a separate PR